gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,189 @@
+ *
+ * <p>This class is not thread-safe, and requires calling classes to make it thread safe.
+ */
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
+    implements Batcher<ElementT, ElementResultT> {
+
+  /** The amount of time to wait before checking responses are received or not. */
+  private static final long DEFAULT_FINISH_WAIT_NANOS = 250_000_000;
+
+  private final BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT>
+      batchingDescriptor;
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final RequestT prototype;
+
+  private final AtomicInteger numOfRpcs = new AtomicInteger(0);
+  private Batch batch;
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = Preconditions.checkNotNull(builder.prototype);
+    this.callable = Preconditions.checkNotNull(builder.unaryCallable);
+    this.batchingDescriptor = Preconditions.checkNotNull(builder.batchingDescriptor);
+  }
+
+  /** Builder for a BatcherImpl. */
+  public static class Builder<ElementT, ElementResultT, RequestT, ResponseT> {
+    private BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor;
+    private UnaryCallable<RequestT, ResponseT> unaryCallable;
+    private RequestT prototype;
+
+    private Builder() {}
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setBatchingDescriptor(
+        BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor) {
+      this.batchingDescriptor = batchingDescriptor;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setUnaryCallable(
+        UnaryCallable<RequestT, ResponseT> unaryCallable) {
+      this.unaryCallable = unaryCallable;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setPrototype(RequestT prototype) {
+      this.prototype = prototype;
+      return this;
+    }
+
+    public BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT> build() {
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(final ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot perform batching on a closed connection");
+
+    if (batch == null) {
+      batch = new Batch(batchingDescriptor.newRequestBuilder(prototype));
+    }
+
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    batch.add(element, result);
+    return result;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    sendBatch();
+    while (numOfRpcs.get() > 0) {
+      LockSupport.parkNanos(DEFAULT_FINISH_WAIT_NANOS);
+    }
```

Comment:
I don't love this. Can we use a semaphore instead?

Reply:
I have added a semaphore implementation 0 permits. Please let me know if this impl looks good  to you 
`private final AtomicInteger numOfRpcs = new AtomicInteger(0);`

Reply:
Thanks for the detailed steps, I have implemented with the same for `Batcher#flush()`. Please have a look.